_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+* **Lack of Support in Modern Ecosystems**: Many modern message buses and RESTful APIs do not support the 2PC protocol, making it challenging to implement in contemporary microservices architectures.
+
+## The Real-World Implications of Naive Approaches
+
+From a theoretical standpoint, the challenges with these naive approaches in integrating an event bus within an application may seem obvious. 
+However, drawing from my professional experience across various projects, I have observed these issues manifest repeatedly in real-world scenarios. 
+Whether it's dealing with distributed systems employing message buses, interacting with external systems through REST APIs, or managing multiple databases, a common error persists: a lack of clarity among developers regarding the consistency semantics of their applications.
+
+In practice, this oversight leads to perplexing and hard-to-diagnose errors. 
+Developers often underestimate the complexity introduced by distributed environments and might not fully consider the implications of their chosen implementation strategy. 
+As a result, applications can exhibit erratic behavior, where the root cause is deeply embedded in the flawed interaction between different system components. 
+These issues can be especially challenging to debug and resolve, given the distributed nature of these systems and the often subtle interdependencies between their components.
+
+The crux of the problem lies in the assumption that techniques and patterns used in monolithic or single-database applications can be directly applied to distributed systems. 
+This misconception overlooks the fundamental differences in how distributed systems handle data consistency and transaction management. 
+The consequences of this oversight are not just theoretical inconsistencies but tangible bugs and system failures that can have a significant impact on the reliability and integrity of the application.
+
+This challenge highlights the inherent difficulty in ensuring atomicity and consistency when dealing with internal transactions and external communication within the same operational scope. 
+It underlines the necessity for a pattern like the Transactional Outbox, which offers a way to decouple these concerns, 
+ensuring that events are only published after the transaction’s successful commitment, thereby maintaining the integrity and consistency of the overall system.
+
+# Understanding the Transactional Outbox
+
+The Transactional Outbox Pattern is a design pattern used in microservices architectures to ensure consistent and reliable message publishing in distributed systems. 
+At its core, the pattern solves a fundamental problem: how to update a database and publish messages/events to a message bus or event-driven system and keep the state consistent.
+
+![transactional-outbox](/assets/article_images/2023-11-17-transactional-outbox/transactional-outbox.png)
+
```

Comment:
possibly another interesting point: how do databases do atomic commits? only relevant if that mechanism is similar to an outbox, but on a lower level

Reply:
This is an implementation detail of the database.

Reply:
will they use a similar mechanism such as outbox, but on a lower level? just curious, do you happen to know how that works?

Reply:
There is a mechanism like the outbox, the translation log, but to be honest I do now have the knowledge about this topic. It is actually hard to find web resources for that topi. I think you have to read some scientific papers from the 80s/90s or a good book. This may be a good reference: https://www.amazon.de/-/en/Database-Internals-Deep-Distributed-Systems/dp/1492040347